cardano-cli/src/Cardano/CLI/Mary/Parser.hs

```diff
@@ -0,0 +1,216 @@
+module Cardano.CLI.Mary.Parser
+  ( ValueExpr (..)
+  , ValueParseError (..)
+  , evalValueExpr
+  , renderValueParseError
+  , stringToValue
+  ) where
+
+import           Prelude
+
+import           Control.Applicative (some, (<|>))
+import           Data.Functor (($>))
+import           Data.Functor.Identity (Identity)
+import           Data.Text (Text, intercalate)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import           Data.Word (Word64)
+import           Text.Parsec (ParseError, parse, try, (<?>))
+import           Text.Parsec.Char (alphaNum, hexDigit, space, spaces, string)
+import           Text.Parsec.Error (errorMessages, messageString)
+import           Text.Parsec.Expr (Assoc (AssocLeft), Operator (Infix, Prefix),
+                     buildExpressionParser)
+import           Text.Parsec.Language (haskellDef)
+import           Text.Parsec.String (Parser)
+import           Text.Parsec.Token (GenTokenParser, decimal, makeTokenParser, parens, reservedOp)
+import           Text.ParserCombinators.Parsec.Combinator (many1, notFollowedBy)
+
+import           Cardano.Api.Typed (AssetId (..), AssetName (..), PolicyId (..), Quantity (..),
+                     ScriptHash (..), Value, selectAsset, valueFromList)
+import           Cardano.Crypto.Hash (hashFromStringAsHex)
+import qualified Shelley.Spec.Ledger.Scripts as Shelley
+
+-- | Parse and construct a 'Value' from its string representation.
+stringToValue :: String -> Either ValueParseError Value
+stringToValue input =
+    either
+      (Left . ValueParseError)
+      (validateValue . evalValueExpr)
+      (parse expression "" input)
+  where
+    validateValue :: Value -> Either ValueParseError Value
+    validateValue v
+      | let ada = selectAsset v AdaAssetId, ada < 0 =
+          Left (NegativeLovelaceValueError ada)
+      | otherwise = Right v
+
+-- | Evaluate a 'ValueExpr' and construct a 'Value'.
+evalValueExpr :: ValueExpr -> Value
+evalValueExpr = go mempty
+  where
+    go :: Value -> ValueExpr -> Value
+    go acc v =
+      case v of
+        ValueExprAdd x y -> acc <> go mempty x <> go mempty y
+        ValueExprLovelace quant -> acc <> valueFromList [(AdaAssetId, quant)]
+        ValueExprMultiAsset polId aName quant ->
+          acc <> valueFromList [(AssetId polId aName , quant)]
+
+textToPolicyId :: Text -> Maybe PolicyId
+textToPolicyId hashText =
+  PolicyId . ScriptHash . Shelley.ScriptHash
+    <$> hashFromStringAsHex (Text.unpack hashText)
```

Comment:
We can use the API directly rather than using other lower level libs and the `ScriptHash` constructor.

The API provides `deserialiseFromRawBytesHex` and `ScriptHash` is a member of `SerialiseAsRawBytes`.